codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessage.java

```diff
@@ -214,7 +232,71 @@ static HAProxyMessage decodeHeader(ByteBuf header) {
             dstPort = header.readUnsignedShort();
         }
 
-        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort);
+        final List<HAProxyTLV> tlvs = readTlvs(header);
+
+        return new HAProxyMessage(ver, cmd, protAndFam, srcAddress, dstAddress, srcPort, dstPort, tlvs);
+    }
+
+    private static List<HAProxyTLV> readTlvs(final ByteBuf header) {
+        HAProxyTLV haProxyTLV = readNextTLV(header);
+        if (haProxyTLV == null) {
+            return Collections.emptyList();
+        }
+        // In most cases there are less than 4 TLVs available
+        List<HAProxyTLV> haProxyTLVs = new ArrayList<HAProxyTLV>(4);
+
+        do {
+            haProxyTLVs.add(haProxyTLV.retain());
```

Comment:
it seems strange that we only `retain()` the "top level" `HAProxyTLV` objects. For example wouldn't you also expect all the `((HAProxySSLTLV) haProxyTLV).encapsulatedTLVs()` added below to be released too? The call to `encapsulatedTlvs.add(haProxyTLV);` below doesn't account for `retain()` either. It may be more strait forward to do a retain (e.g `readRetainedSlice()` or `retainedSlice()`) in `readNextTLV` when ever we get the "raw content" (for `HAProxySSLTLV` and `HAProxyTLV`)